[jianminglok] iP

jianminglok opened the pull request:
# TwoFive
### Your day-to-day task companion

TwoFive makes it **easier** for you to remind your daily tasks. It's
- Text-based/CLI based 🗒️ 
- User friendly 💯 
- Responds to your queries **fast** 🚀

It's easy to get started with using TwoFive, just follow the steps below.
1. Download it from [here](https://github.com/jianminglok/ip/releases/download/A-Jar/TwoFive.jar)
2. Double click the file named TwoFive.jar that you have just downloaded
3. Add your tasks and manage them!

Here's a list of features:
- [X] Managing tasks
- [X] Managing events
- [X] Managing deadlines

Commit:
Add Gradle support

jianminglok committed:
Changed name for chatbot and added greetings

jianminglok committed:
Reads input from user and echo it

jianminglok committed:
Added Task Class

jianminglok committed:
Store and list tasks entered by user

jianminglok committed:
Rephrased task name as description and added task status

jianminglok committed:
Mark tasks as done or not done

jianminglok committed:
Add ToDo, Deadline & Event tasks

jianminglok committed:
Add different type of tasks through different commands

jianminglok committed:
Format changes

jianminglok committed:
Added semi-auto testing for TwoFive

jianminglok committed:
Add error handling through exceptions

jianminglok committed:
Updated error format and added tests for error handling

jianminglok committed:
Add ability to delete tasks

jianminglok committed:
Added more tests for delete, mark and unmark feature

jianminglok committed:
Add ability to save and load task file

jianminglok committed:
Add tests for loading and saving task file

jianminglok committed:
Allow TwoFive to understand dates and times

jianminglok committed:
Add command to list deadlines/events on a specific date

jianminglok committed:
Add tests for date and time update

jianminglok committed:
Merge branch 'branch-Level-7'

jianminglok committed:
Resolve merge conflicts

jianminglok committed:
Resolve merge conflicts and merge branch 'branch-Level-8'

jianminglok committed:
Merge branch 'master' of https://github.com/jianminglok/ip

jianminglok committed:
Extract classes for better OOP adherance

jianminglok committed:
Add more OOP updates

jianminglok committed:
Divide classes into packages

jianminglok committed:
Update automated testing script

jianminglok committed:
Merge remote-tracking branch 'origin/add-gradle-support'

jianminglok committed:
Add gradle support

jianminglok committed:
Override equals method for Task and its subclasses

jianminglok committed:
Add JUnit tests

jianminglok committed:
Update build.gradle file

jianminglok committed:
Add JavaDoc comments

jianminglok committed:
Comply with coding standards

jianminglok committed:
Add find command for users to filter task with a specified keyword

jianminglok committed:
Add tests for find command

jianminglok committed:
Follow coding standard for unit test

jianminglok committed:
Merge branch 'branch-A-JavaDoc'

jianminglok committed:
Resolve merge conflicts

jianminglok committed:
Resolve merge conflicts for 'branch-Level-9'

jianminglok committed:
Add missed JavaDoc

jianminglok committed:
Add gradle checkstyle

jianminglok committed:
Update code according to Gradle checkstyle

jianminglok committed:
Update test classes according to Gradle checkstyle

jianminglok committed:
Save tasks to file instantly when changes made

gohyongjing commented on src/main/java/twofive/utils/FileParser.java:
```diff
@@ -0,0 +1,123 @@
+ * Parses the contents of a given file into a list of tasks.
+ */
+public class FileParser {
+    private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
```
Perhaps you could consider changing formatter to be in full caps since it is constant class attribute

gohyongjing commented on src/main/java/twofive/command/DeleteCommand.java:
```diff
@@ -0,0 +1,44 @@
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws InvalidTaskException {
+        if (this.taskNum < 0 || this.taskNum >= tasks.getTasksNum()) {
```
Perhaps you could avoid the use of this keyword to reduce unecessary noise
```suggestion
        if (taskNum < 0 || taskNum >= tasks.getTasksNum()) {
```

gohyongjing commented on src/main/java/twofive/command/FindCommand.java:
```diff
@@ -0,0 +1,28 @@
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        ui.showMessage(tasks.getKeywordString(this.keyword));
```
Perhaps you could remove the this keyword here too to reduce unecessary noise, as well as other parts of the code too
```suggestion
        ui.showMessage(tasks.getKeywordString(keyword));
```

gohyongjing commented on src/main/java/twofive/utils/Parser.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    // Ensure a valid description is provided when adding a new task
+    private static String validateDescription(String input, String commandWord) throws EmptyDescriptionException {
```
Perhaps you could consider writing javadocs for this method as it appears to do a non-trivial task

gohyongjing reviewed (commented):
I think your code is very well written and follows major coding standards, except for a few small details under advanced rules of the given coding standards 👍

jianminglok committed:
Add GUI for TwoFive

jianminglok committed:
Update existing code for GUI implementation

jianminglok committed:
Update tests

jianminglok commented:
@gohyongjing Thanks for pointing out these details even though they were small, they proved very helpful indeed!

jianminglok committed:
Update code according to coding standard

jianminglok committed:
Merge branch 'branch-Level-10'

arkobrat commented on src/main/java/twofive/command/Command.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Results in various actions being performed when commands are executed
+ * in initialized subclasses of this class.
+ */
```
Great job going above and beyond the minimal requirements in adding separate classes for Commands

arkobrat commented on src/main/java/twofive/exception/EmptyDateException.java:
```diff
@@ -0,0 +1,12 @@
+/**
+ * Represents an error encountered when a date is not provided for parsing
+ * in a DueCommand.
+ */
```
Extensive handling of exceptions, which is hallmark of good code but consider adding an overarching DukeException class to pass exceptions to?

arkobrat commented on src/main/java/twofive/TwoFive.java:
```diff
@@ -0,0 +1,60 @@
+     *
+     * @param filePath String containing file path.
+     */
+    public TwoFive(String filePath) throws IOException, TwoFiveException, DateTimeParseException {
```
```suggestion
    public TwoFive(String filePath) throws DukeException {
```

Consider changing your exception handling to just handle DukeException which could be the umbrella exception class the other exceptions extend from

jianminglok replied:
Thank you for the suggestion and I greatly appreciate the other comments made. There is currently an umbrella exception named TwoFiveException instead of DukeException, but what you pointed out is still a good practice to keep it mind. 👍

arkobrat commented on src/main/java/twofive/data/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+package twofive.data;
```
Great use of packaging beyond the required single-package usage

arkobrat reviewed (commented):
Great code overall, the extensive exception handling was great to see as was the division of related work into separate classes and the very well-done packaging.

jianminglok committed:
Add assertions

Assertions are previously missing so assumptions in the code cannot
be verified.

Assertions are now added so that possible bugs can be identified.

jianminglok committed:
Update code according to checkstyle

jianminglok committed:
Update code by following code quality guidelines

Some code written did not follow the proper code guideline, so
improvements to the code quality are made.

This helps to improve readability and make the structure more
logical.

jianminglok committed:
Update code according to checkstyle after improving code quality

jianminglok committed:
Use Streams in the code

Streams have been added for usage in the task list.

The usage of Streams helps to improve readability through a higher
level of abstraction.

jianminglok committed:
Merge pull request #1 from jianminglok/branch-A-Assertions

jianminglok committed:
Resolve merge conflicts

jianminglok committed:
Merge pull request #2 from jianminglok/branch-A-CodeQuality

Branch A-CodeQuality

jianminglok committed:
Merge branch 'master' into branch-A-Streams

jianminglok committed:
Merge pull request #3 from jianminglok/branch-A-Streams

Use Streams in the code

jianminglok committed:
Set up CI

jianminglok committed:
Add .yml file for CI

jianminglok committed:
Update unresolved merge conflicts

jianminglok committed:
Update test according to gradle check

jianminglok committed:
Add ability to tag tasks

jianminglok committed:
Merge branch 'branch-C-Tagging'

jianminglok committed:
Add a better GUI to improve user experience

jianminglok committed:
Give TwoFive a unique personality

jianminglok committed:
Merge branch 'branch-A-BetterGui'

jianminglok committed:
Merge branch 'branch-A-Personality'

jianminglok committed:
Add product screenshot

jianminglok committed:
Add user guide

jianminglok committed:
Fix user guide

jianminglok committed:
Update user guide

jianminglok committed:
Update README for macOS users
